src/backends/policy_engine/souffle/datalog_lowering_visitor.h

```diff
@@ -0,0 +1,118 @@
+#ifndef SRC_BACKENDS_POLICY_ENGINE_SOUFFLE_DATALOG_LOWERING_VISITOR_H_
+#define SRC_BACKENDS_POLICY_ENGINE_SOUFFLE_DATALOG_LOWERING_VISITOR_H_
+
+#include "src/common/logging/logging.h"
+#include "src/ir/attributes/attribute.h"
+#include "src/ir/attributes/int_attribute.h"
+#include "src/ir/attributes/string_attribute.h"
+#include "src/ir/datalog/operation.h"
+#include "src/ir/ir_visitor.h"
+#include "src/ir/module.h"
+#include "src/ir/operator.h"
+#include "src/ir/ir_traversing_visitor.h"
+
+namespace raksha::backends::policy_engine::souffle {
+
+// A class containing the Datalog facts produced by the IR translator in
+// structured (ie, C++ objects, not strings) form.
+class RakshaDatalogFacts {
+ public:
+  void AddIsOperationFact(ir::datalog::IsOperationFact fact) {
+    is_operation_facts_.push_back(std::move(fact));
+  }
+
+  std::string ToDatalogString() const {
+    return absl::StrJoin(
+        is_operation_facts_, "\n",
+        [](std::string *out, const ir::datalog::IsOperationFact &arg) {
+          absl::StrAppend(out, arg.ToDatalogString());
+        });
+  }
+
+ private:
+  std::vector<ir::datalog::IsOperationFact> is_operation_facts_;
+};
+
+ class DatalogLoweringVisitor : public ir::IRTraversingVisitor<DatalogLoweringVisitor> {
+ public:
+  // TODO: Dedup this from `DecoderContext` in the SQL frontend.
+  static constexpr absl::string_view kDefaultOutputName = "out";
+  DatalogLoweringVisitor(ir::SsaNames &ssa_names) : ssa_names_(ssa_names) {}
+  virtual ~DatalogLoweringVisitor() {}
+
+  void PreVisit(const ir::Operation &operation) override {
+    // We currently don't have any owner information when outputting IR. We
+    // don't need it yet, really, but we do need to output something.
+    constexpr absl::string_view kUnknownPrincipal = "UNKNOWN";
+    const ir::Operator &op = operation.op();
+    absl::string_view op_name = op.name();
+
+    const ir::ValueList &ir_operand_list = operation.inputs();
+    ir::datalog::OperandList operand_list =
+        RangeToDatalogList<ir::datalog::OperandList>(ir_operand_list.begin(),
```

Comment:
You could use `absl::c_accumulate` instead of creating a new method? I think that will be more readable?
(https://en.cppreference.com/w/cpp/algorithm/accumulate)

Reply:
`c_accumulate` would definitely be more readable, I hadn't considered that. The one issue is that the documentation for `std::accumulate` indicates that the accumulate operation should have no side effects. If `c_accumulate` has a similar constraint, we can't actually use it, as we alter `ssa_names_` in one of the called functions. I'll check it out.

Reply:
It looks like `c_accumulate` is just a wrapper around `std::accumulate`, so I don't think I can use it (until C++20).

Reply:
I think it is available before C++20. It is just that some API is changing.

Reply:
Ah, you are saying that you can't use because of side effects?

Reply:
Right. The guidance from [C++11, C++20) is that the `BinaryOperator` used within `std::accumulate` cannot have side effects. That's unusually strong guidance about side effects for C++ and I am a bit anxious about crossing that specified line.

Reply:
I must have misread this; `std::accumulate` allows side effects from C++11 on. That means I can use it and it might significantly reduce the code needed here.

Reply:
OK, there is another very annoying reason that `std::accumulate` won't work here. As of C++20, `std::accumulate` uses `std::move` to move the accumulated value from one "iteration" of the fold to another. Before that, it uses a copy. Because `OperandList` and `AttributeList` both contain `std::unique_ptr`s, we can't use `std::accumulate` for this (until C++20).

Reply:
Added a `fold` in #539 . Depending on the PR's respective timelines, will rewrite in terms of `fold` in this or a followup PR.

Reply:
Fixed.